cEP-0033.md

```diff
@@ -0,0 +1,760 @@
+		"arg_dict": arg_dict
+	}
+
+	>>> nl_sections
+	(	<NlSection object(index=1, language=html, orig_start=1, orig_end=3)>,
+		<NlSection object(index=2, language=php, orig_start=4, orig_end=6)>,
+		<NlSection object(index=3, language=html, orig_start=7, orig_end=7)>
+	)
+
+	>>> file_contents = get_file(nl_file_info['file_name'])
+
+	>>> file_contents = \
+			"""
+			<!DOCTYPE html>
+ 			<head>
+			<title>Hello world as text</title>
+			<?php
+				echo "<p>Hello world</p>";
+			?>
+			</html>
+			"""
+
+	>>> nl_file_dict = get_nl_file_dict(nl_file_info, nl_sections)
+	>>> nl_file_dict
+	{
+		"temp_html_file":("<!DOCTYPE html>\n
+				   <head>\n
+				   <title>Hello world as text</title>\n
+				   \n
+				   \n		
+				   \n
+				   </html>"),
+
+		"temp_php_file":('\n
+				  \n
+				  \n
+				  \t\t<?php\n
+				  \t\t\techo "<p>Hello world</p>";\n
+				  \t?>')
+
+
+	}
+
+	'''
+
+def get_file_contents(file_name):
+	""" Returns the contents of the file"""
+
+def assemble(nl_diff_dict, nl_sections, nl_info_dict):
+	"""
+	Assembles the temporary files. 
+
+	The sections are extracted by their increasing order of their index and 
+	then written directly to the original file.
+	"""
+	file = open(nl_info_dict['absolute_file_path'], 'w')
+
+	for nl_section in sorted(nl_sections.index):
+		temp_file_name = nl_section.temp_file
+		start = nl_section.linted_start
+		end = nl_section.linted_end
+		linted_file_content = nl_diff_dict[temp_file_name]
+
+		section_content = get_file_content(linted_file_content, start, end)
+
+		file.write(section_content)
+		"""
+
+```
+
+## Phase 3: Linting the file
+
+This phase deals with linting the temporary files created by the NlFileHandler using
+the `nl_sections`.
+
+1. In coala the linting of the files are done when the section created by coala
+are executed.
+
+2. Since we are explicitly making the sections for nested languages, care has been 
+taken to keep the file names of the `nl.lang` section and the file names
+in `nl_file_dict`  same.
+
+3. During the execution of the section, in the `instantiate_process()` we do not
+access the physical file, rather we replace the `file_dict` with `nl_file_dict`
+and make the necessary changes.
+
+4. Once the `file_dict` is chnaged, coala will normally continue the process.
+To coala, it now looks as if the `temporary files` were actually present in the
+physical drive. And the linting starts
+
+### Applying the Patches
+
+Whenever the bears suggests a patch and the user desires to apply the Patch, we
+would also need to update the information of the `linted_start` and 
+`linted_end`. This needs to be done because, whenever a patch is applied the 
+position of the lines might change because of addition and deletion of lines.
+
+Keeping track of the start and end of a particular `nl_section` in the linted
+files would help in easier extraction of the `nl_section` from the temporary
+created files.
+
+In order to do so, we'll have to make changes to the `apply()` method of
```

Comment:
Do we really need to fiddle with this?   
Looks a little messy. Would be just wonderful if we can instead let the apply function do it's linting in the in-memory sections and while combining them we do some sort of post-processing and instead of modifying the `apply` function, just take care in the assemble phase?

Reply:
We won't be changing the functionality of  `apply()` function. The only thing we would
be doing, is we would ask it to run `update_nl_sections()` present in `NlCore.py ` before applying the patch, so that the `linted_start` and `linted_end` values of the `nl_section` get updated.

Without keeping track of the start and end of the section in the linted file, we won't be able to split the  temporary linted file into sections, as the file might have been updated.

Reply:
What I meant was minimal changes, e.g a decorator of some sort so apply stays as is, and we do whatever we need to before and after apply is called
But yeah if there's only a function call that needs to be added then it wouldn't make much of a difference I guess

Reply:
According to the plan as of now, it's only a function. But later during implementation, If I do find out
that we might need more the a function call - I surely keep your advice in mind and implement them accordingly ^_^